mlflow/store/artifact/databricks_artifact_repo.py

```diff
@@ -201,26 +244,30 @@ def _azure_upload_file(self, credentials, local_file, artifact_path):
         """
         try:
             headers = self._extract_headers_from_credentials(credentials.headers)
-            uploading_block_list = []
-            for chunk in yield_file_in_chunks(local_file, _AZURE_MAX_BLOCK_CHUNK_SIZE):
-                # Base64-encode a UUID, producing a UTF8-encoded bytestring. Then, decode
-                # the bytestring for compliance with Azure Blob Storage API requests
-                block_id = base64.b64encode(uuid.uuid4().hex.encode()).decode("utf-8")
-                try:
-                    put_block(credentials.signed_uri, block_id, chunk, headers=headers)
-                except requests.HTTPError as e:
-                    if e.response.status_code in [401, 403]:
-                        _logger.info(
-                            "Failed to authorize request, possibly due to credential expiration."
-                            " Refreshing credentials and trying again..."
-                        )
-                        credential_info = self._get_write_credential_infos(
-                            run_id=self.run_id, paths=[artifact_path]
-                        )[0]
-                        put_block(credential_info.signed_uri, block_id, chunk, headers=headers)
-                    else:
-                        raise e
-                uploading_block_list.append(block_id)
+            futures = {}
+            num_chunks = math.ceil(os.path.getsize(local_file) / _AZURE_MAX_BLOCK_CHUNK_SIZE)
+            with self.thread_pool as executor:
+                for index in range(num_chunks):
+                    start_byte = index * _AZURE_MAX_BLOCK_CHUNK_SIZE
+                    upload_future = executor.submit(
+                        self._azure_upload_chunk,
+                        credentials=credentials,
+                        headers=headers,
+                        local_file=local_file,
+                        artifact_path=artifact_path,
+                        start_byte=start_byte,
+                    )
+                    futures[upload_future] = index
+
+                results, errors = self._complete_futures(futures)
+
+            if errors:
```

Comment:
Is it possible to have some level of optimistic retries on errors? 
If we were to do some fixed-iteration retry, we could always reuse the threadpool (switch from context-based pool instantiation to manually creating the pool, submitting failures back to it, and then shutting down the pool at the end of the retry loop). 
I would think that there are a few HTTPErrors that would be flagged as "do not retry" such as authentication failures, but for packet loss errors or checksum failures, a retry might be a relatively simple and very useful thing to implement for this.

Reply:
@BenWilson2 Thanks for the comment!

`cloud_storage_http_request` makes a request to upload a part and this function supports automatic retry:

https://github.com/mlflow/mlflow/blob/d55451f3e9ffd4c6f6eec325a581c5a7d81851a6/mlflow/utils/rest_utils.py#L299